### PR TITLE
Adds a CBMC proof for aws_cryptosdk_hdr_size function

### PR DESCRIPTION
**Pull request by tegansb**
*Description of changes:*
Proof harness and Makefile for aws_cryptosdk_hdr_size. 
Adds new validity, is_bounded and ensures function for the aws_cryptosdk_hdr struct. 
Adds preconditions to the aws_cryptosdk_hdr_size function. 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

# Check any applicable:
- [ ] Were any files moved? Moving files changes their URL, which breaks all hyperlinks to the files.

### CODE REVIEW

File: include/aws/cryptosdk/private/header.h [hunk 1]
```diff
@@ -126,4 +126,9 @@ bool aws_cryptosdk_algorithm_is_known(uint16_t alg_id);
  */
 bool aws_cryptosdk_algorithm_is_committing(uint16_t alg_id);
 
+/**
+ * Returns true if aws_cryptosdk_hdr is valid and false otherwise.
+ */
```
Comment (feliperodri) [inline include/aws/cryptosdk/private/header.h:131]: I'd suggest
```c
/**
 * Evaluates the set of properties that define the shape of all valid aws_cryptosdk_hdr structures.
 * It is also a cheap check, in the sense it run in constant time (i.e., no loops or recursion).
 */
```

File: source/header.c [hunk 2]
```diff
@@ -456,3 +457,11 @@ int aws_cryptosdk_hdr_write(
     *bytes_written = 0;
     return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
 }
+
+bool aws_cryptosdk_hdr_is_valid(const struct aws_cryptosdk_hdr *hdr) {
+    /* header is not NULL and each field is valid */
```
Comment (feliperodri) [inline source/header.c:462]:
```c
    /* Header is not NULL and each field is a valid structure. */
```

File: verification/cbmc/include/aws/cryptosdk/private/config.h [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#pragma once
+
+/* aws-encryption-sdk-crequires a configuration file.
+ * We do not currently need to set any options.
+ */
```
Comment (feliperodri) [inline verification/cbmc/include/aws/cryptosdk/private/config.h:20]:
```c
/* 
 * AWS Encryption SDK for C requires a configuration file.
 * We do not currently need to set any options for CBMC proofs.
 */
```

File: verification/cbmc/include/make_common_data_structures.h [hunk 4]
```diff
@@ -38,6 +39,14 @@ bool aws_cryptosdk_edk_list_elements_are_bounded(const struct aws_array_list *co
 void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list);
 void ensure_cryptosdk_edk_list_has_allocated_list_elements(struct aws_array_list *list);
 
+/* Allocates the members of the header and ensures that internal pointers are pointing to the correct objects. */
+struct aws_cryptosdk_hdr *ensure_nondet_hdr_has_allocated_members();
+/* Determines if the members of the header are bounded. */
+bool aws_cryptosdk_hdr_members_are_bounded(
+    const struct aws_cryptosdk_hdr *const hdr, const size_t max_edk_item_size, const size_t max_item_size);
```
Comment (feliperodri) [inline verification/cbmc/include/make_common_data_structures.h:46]: You don't need to use `const` in pointers, so `const struct aws_cryptosdk_hdr *hdr` is enough.

File: verification/cbmc/include/make_common_data_structures.h [hunk 5]
```diff
@@ -38,6 +39,14 @@ bool aws_cryptosdk_edk_list_elements_are_bounded(const struct aws_array_list *co
 void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list);
 void ensure_cryptosdk_edk_list_has_allocated_list_elements(struct aws_array_list *list);
 
+/* Allocates the members of the header and ensures that internal pointers are pointing to the correct objects. */
+struct aws_cryptosdk_hdr *ensure_nondet_hdr_has_allocated_members();
+/* Determines if the members of the header are bounded. */
```
Comment (feliperodri) [inline verification/cbmc/include/make_common_data_structures.h:44]: Add an extra empty line among declarations.

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+HARNESS_ENTRY = $(PROOF_UID)_harness
+HARNESS_FILE = $(HARNESS_ENTRY).c
+
+MAX_EDK_LIST_ITEMS ?= 2
+MAX_TABLE_SIZE ?= 2
+
+DEFINES += -DAWS_NO_STATIC_IMPL
+DEFINES += -DMAX_EDK_LIST_ITEMS=$(MAX_EDK_LIST_ITEMS)
+DEFINES += -DMAX_TABLE_SIZE=$(MAX_TABLE_SIZE)
+DEFINES += -DMEMCPY_STUB_MAX=128
```
Comment (feliperodri) [inline verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile:31]: Why 128? Add a comment.

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+include ../Makefile.local_default
+include ../Makefile.aws_byte_buf
+
+PROOF_UID = aws_cryptosdk_hdr_size
+
+HARNESS_ENTRY = $(PROOF_UID)_harness
+HARNESS_FILE = $(HARNESS_ENTRY).c
+
+MAX_EDK_LIST_ITEMS ?= 2
+MAX_TABLE_SIZE ?= 2
```
Comment (feliperodri) [inline verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile:26]: Why 2? Add a comment.

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+PROOF_SOURCES += $(PROOFDIR)/$(HARNESS_FILE)
+
+REMOVE_FUNCTION_BODY += aws_cryptosdk_enc_ctx_size
+REMOVE_FUNCTION_BODY += aws_array_list_get_at_ptr
+
+UNWINDSET += aws_cryptosdk_edk_list_elements_are_bounded.0:$(call addone,$(MAX_EDK_LIST_ITEMS))
+UNWINDSET += aws_cryptosdk_edk_list_elements_are_valid.0:$(call addone,$(MAX_EDK_LIST_ITEMS))
+UNWINDSET += aws_cryptosdk_hdr_size.0:$(call addone,$(MAX_EDK_LIST_ITEMS)) 
+UNWINDSET += ensure_cryptosdk_edk_list_has_allocated_list_elements.0:$(call addone,$(MAX_EDK_LIST_ITEMS))
+UNWINDSET += memcmp.0:$(shell echo $$(($(MAX_BUFFER_SIZE) + 1)))
```
Comment (feliperodri) [inline verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile:67]: You can use `addone` here.

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/aws_cryptosdk_hdr_size_harness.c [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+    __CPROVER_assume(hdr->enc_ctx.p_impl != NULL);
+    ensure_hash_table_has_valid_destroy_functions(&hdr->enc_ctx);
+    size_t empty_slot_idx;
+    __CPROVER_assume(aws_hash_table_has_an_empty_slot(&hdr->enc_ctx, &empty_slot_idx));
+
+    /* Operation under verification */
+    aws_cryptosdk_hdr_size(hdr);
+
+    /* Postconditions */
+    assert(aws_cryptosdk_hdr_is_valid(hdr));
```
Comment (feliperodri) [inline verification/cbmc/proofs/aws_cryptosdk_hdr_size/aws_cryptosdk_hdr_size_harness.c:42]: Why about immutability? The structure cannot change after this operation.

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/cbmc-batch.yaml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+build_memory: 32000
+cbmcflags: "--unwind;1;--unwindset;aws_cryptosdk_edk_list_elements_are_bounded.0:3,aws_cryptosdk_edk_list_elements_are_valid.0:3,aws_cryptosdk_hdr_size.0:3,ensure_cryptosdk_edk_list_has_allocated_list_elements.0:3,memcmp.0:1029;--flush;--object-bits;8;--external-sat-solver;kissat;--malloc-may-fail;--malloc-fail-null;--bounds-check;--conversion-check;--div-by-zero-check;--float-overflow-check;--nan-check;--pointer-check;--pointer-overflow-check;--pointer-primitive-check;--signed-overflow-check;--undefined-shift-check;--unsigned-overflow-check;--unwinding-assertions;--malloc-may-fail;--malloc-fail-null"
+coverage_memory: 32000
+expected: SUCCESSFUL
+goto: /Users/tegab/clean/aws-encryption-sdk-c/verification/cbmc/proofs/aws_cryptosdk_hdr_size/gotos/aws_cryptosdk_hdr_size_harness.c.goto
+jobos: ubuntu16
+property_memory: 32000
+report_memory: 32000
```
Comment (feliperodri) [inline verification/cbmc/proofs/aws_cryptosdk_hdr_size/cbmc-batch.yaml:8]: I don't think we need this `cbmc-batch.yaml` file anymore, you can replace it with the standard one.
```Makefile
:
  This file marks this directory as containing a CBMC proof. This file
  is automatically clobbered in CI and replaced with parameters for
  running the proof.
```

File: verification/cbmc/sources/make_common_data_structures.c [hunk 11]
```diff
@@ -126,6 +127,9 @@ bool aws_cryptosdk_edk_list_elements_are_bounded(const struct aws_array_list *co
 }
 
 void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list) {
+    if (list == NULL) {
+        return;
+    }
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:132]: I'd recommend
```c
if (list != NULL) {
    if (list->current_size == 0) {
        /* ... */
    } else {
        /* ... */
    }
}
```

File: verification/cbmc/sources/make_common_data_structures.c [hunk 12]
```diff
@@ -137,12 +141,82 @@ void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list) {
 }
 
 void ensure_cryptosdk_edk_list_has_allocated_list_elements(struct aws_array_list *list) {
+    if (!aws_cryptosdk_edk_list_is_valid(list)) {
+        return;
+    }
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:146]: Same as above.

File: verification/cbmc/sources/make_common_data_structures.c [hunk 13]
```diff
@@ -137,12 +141,82 @@ void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list) {
...
+    struct aws_hash_table *enc_ctx = malloc(sizeof(*enc_ctx));
+    if (enc_ctx) {
+        hdr->enc_ctx = *enc_ctx;
+    }
+    struct aws_array_list *edk_list = malloc(sizeof(*edk_list));
+    if (edk_list) {
+        hdr->edk_list = *edk_list;
+    }
+
+    hdr->auth_len = malloc(sizeof(hdr->auth_len));
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:188]: I'd recommend
```c
if (hdr != NULL) {
    hdr->alloc = nondet_bool() ? NULL : can_fail_allocator();
    hdr->frame_len = malloc(sizeof(hdr->frame_len));
    ensure_byte_buf_has_allocated_buffer_member(&hdr->iv);
    ensure_byte_buf_has_allocated_buffer_member(&hdr->auth_tag);
    ensure_byte_buf_has_allocated_buffer_member(&hdr->message_id);
    ensure_byte_buf_has_allocated_buffer_member(&hdr->alg_suite_data);
    ensure_allocated_hash_table(&hdr->enc_ctx, /* YOU NEED A BOUND HERE. */);
    ensure_array_list_has_allocated_data_member(&hdr->edk_list);
}
```
If a structure field is not a pointer, you don't need to allocate or make sure is non-deterministic, CBMC will do that automatically. You need to make sure the pointer structures are properly allocated, and I'd suggest using the existing ensure functions.

File: verification/cbmc/sources/make_common_data_structures.c [hunk 14]
```diff
@@ -137,12 +141,82 @@ void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list) {
...
+
+    return hdr;
+}
+
+bool aws_cryptosdk_hdr_members_are_bounded(
+    const struct aws_cryptosdk_hdr *const hdr, const size_t max_edk_item_size, const size_t max_item_size) {
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:194]: No need for `const` pointers.

File: verification/cbmc/sources/make_common_data_structures.c [hunk 15]
```diff
@@ -137,12 +141,82 @@ void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list) {
...
+           aws_byte_buf_is_bounded(&hdr->message_id, max_item_size) &&
+           aws_byte_buf_is_bounded(&hdr->iv, max_item_size) &&
+           aws_byte_buf_is_bounded(&hdr->alg_suite_data, max_item_size) &&
+           aws_byte_buf_is_bounded(&hdr->auth_tag, max_item_size);
+}
+
+void ensure_hdr_members_have_allocated_members(struct aws_cryptosdk_hdr *hdr, const size_t max_table_size) {
+    if (hdr == NULL) {
+        return;
+    }
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:211]: Wrap the entire implementation with it.

File: verification/cbmc/sources/make_common_data_structures.c [hunk 16]
```diff
@@ -137,12 +141,82 @@ void ensure_cryptosdk_edk_list_has_allocated_list(struct aws_array_list *list) {
...
+    ensure_cryptosdk_edk_list_has_allocated_list(&hdr->edk_list);
+    ensure_byte_buf_has_allocated_buffer_member(&hdr->message_id);
+    ensure_byte_buf_has_allocated_buffer_member(&hdr->iv);
+    ensure_byte_buf_has_allocated_buffer_member(&hdr->alg_suite_data);
+    ensure_byte_buf_has_allocated_buffer_member(&hdr->auth_tag);
+    ensure_allocated_hash_table(&hdr->enc_ctx, max_table_size);
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:217]: Nice!!!

**Review by feliperodri — CHANGES_REQUESTED** [review]

File: include/aws/cryptosdk/private/header.h (hunk 1 above)
Comment (tegansb) [reply include/aws/cryptosdk/private/header.h:131]: Updated as suggested!

File: source/header.c (hunk 2 above)
Comment (tegansb) [reply source/header.c:462]: Updated!

File: verification/cbmc/include/aws/cryptosdk/private/config.h (hunk 3 above)
Comment (tegansb) [reply verification/cbmc/include/aws/cryptosdk/private/config.h:20]: Updated!

File: verification/cbmc/include/make_common_data_structures.h (hunk 4 above)
Comment (tegansb) [reply verification/cbmc/include/make_common_data_structures.h:46]: Fixed!

File: verification/cbmc/include/make_common_data_structures.h (hunk 5 above)
Comment (tegansb) [reply verification/cbmc/include/make_common_data_structures.h:44]: Updated!

File: verification/cbmc/sources/make_common_data_structures.c (hunk 11 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c:132]: Updated!

File: verification/cbmc/sources/make_common_data_structures.c (hunk 12 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c:146]: Also updated.

File: verification/cbmc/sources/make_common_data_structures.c (hunk 13 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c:188]: Oh I see! Updated as suggested!

File: verification/cbmc/sources/make_common_data_structures.c (hunk 14 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c:194]: Fixed!

File: verification/cbmc/sources/make_common_data_structures.c (hunk 15 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c:211]: Done!

File: verification/cbmc/sources/make_common_data_structures.c (hunk 16 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c:217]: Thanks!

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/cbmc-batch.yaml (hunk 10 above)
Comment (tegansb) [reply verification/cbmc/proofs/aws_cryptosdk_hdr_size/cbmc-batch.yaml:8]: Updated!

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile (hunk 6 above)
Comment (tegansb) [reply verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile:31]: Just experimented and seems this DEFINES is un-needed and removing it doesn't really impact performance, so I've taken it out.

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile (hunk 7 above)
Comment (tegansb) [reply verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile:26]: Added comments for both.

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile (hunk 8 above)
Comment (tegansb) [reply verification/cbmc/proofs/aws_cryptosdk_hdr_size/Makefile:67]: Done!

File: verification/cbmc/proofs/aws_cryptosdk_hdr_size/aws_cryptosdk_hdr_size_harness.c (hunk 9 above)
Comment (tegansb) [reply verification/cbmc/proofs/aws_cryptosdk_hdr_size/aws_cryptosdk_hdr_size_harness.c:61]: Added immutability check. I wanted to wrap all the checks in a single function but having to keep all the store_byte_from_buffer made it simpler to check each aws_byte_buf individually.

File: source/edk.c [hunk 17]
```diff
@@ -72,13 +72,10 @@ bool aws_cryptosdk_edk_is_valid(const struct aws_cryptosdk_edk *const edk) {
...
-    for (size_t i = 0; i < len; ++i) {
-        struct aws_cryptosdk_edk *edk;
-        if (!aws_array_list_get_at_ptr(edk_list, (void **)&edk, i)) {
-            if (!aws_cryptosdk_edk_is_valid(edk)) {
-                return false;
-            }
+    for (size_t i = 0; i < edk_list->length; ++i) {
+        struct aws_cryptosdk_edk *data = (struct aws_cryptosdk_edk *)edk_list->data;
+        if (!aws_cryptosdk_edk_is_valid(&(data[i]))) {
+            return false;
```
Comment (alex-chew) [inline source/edk.c:78]: Is this change just a verification performance optimization?

File: verification/cbmc/sources/make_common_data_structures.c [hunk 18]
```diff
@@ -126,21 +127,55 @@ bool aws_cryptosdk_edk_list_elements_are_bounded(const struct aws_array_list *co
...
+    const struct aws_cryptosdk_hdr *hdr, const size_t max_edk_item_size, const size_t max_item_size) {
+    if (hdr != NULL) {
+        /*IV buffer length might need further constraints, this is done in the harness when necessary */
+        return aws_cryptosdk_edk_list_is_bounded(&hdr->edk_list, max_edk_item_size) &&
+               (!aws_cryptosdk_edk_list_is_valid(&hdr->edk_list) ||
+                aws_cryptosdk_edk_list_elements_are_bounded(&hdr->edk_list, max_item_size)) &&
+               aws_byte_buf_is_bounded(&hdr->message_id, max_item_size) &&
+               aws_byte_buf_is_bounded(&hdr->iv, max_item_size) &&
+               aws_byte_buf_is_bounded(&hdr->alg_suite_data, max_item_size) &&
+               aws_byte_buf_is_bounded(&hdr->auth_tag, max_item_size);
```
Comment (alex-chew) [inline verification/cbmc/sources/make_common_data_structures.c:176]: nitpick: It'll be easier to maintain these `byte_buf`-related lines if they have the same order whenever they appear, e.g. in order of declaration within the header struct

**Review by alex-chew — APPROVED** [review]
Overall LGTM, just a couple minor comments/questions

File: verification/cbmc/sources/make_common_data_structures.c [hunk 19]
```diff
@@ -126,21 +127,55 @@ bool aws_cryptosdk_edk_list_elements_are_bounded(const struct aws_array_list *co
...
+            ensure_cryptosdk_edk_has_allocated_members(&(data[i]));
+        }
+    }
+}
+
+struct aws_cryptosdk_hdr *ensure_nondet_hdr_has_allocated_members(const size_t max_table_size) {
+    struct aws_cryptosdk_hdr *hdr = malloc(sizeof(*hdr));
+    if (hdr != NULL) {
+        hdr->alloc     = nondet_bool() ? NULL : can_fail_allocator();
+        hdr->frame_len = malloc(sizeof(hdr->frame_len));
```
Comment (feliperodri) [inline verification/cbmc/sources/make_common_data_structures.c:155]: if `frame_len` is not a pointer, you don't need that. https://github.com/aws/aws-encryption-sdk-c/blob/fc974765d986662589e8c9f1a07eae6e9bf83665/include/aws/cryptosdk/private/header.h#L29

**Review by feliperodri — APPROVED** [review]
Approved module a few nitpicks. @tegansb

**Comment by alex-chew** [comment]
I think the one failing test in the CodeBuild csdk-ubuntu-latest-x64 build is due to the uninitialized memory in the test, which we fixed in master. We should rebase this onto master and then ensure all the builds pass before merging.

File: source/edk.c (hunk 17 above)
Comment (tegansb) [reply source/edk.c:78]: I was running into a CBMC error with the old implementation, so I changed it to this to match what we do in make_common_data_structures.c. I believe the actual CBMC error was caused elsewhere, but left the change in to be consistent with how we handle edk_lists elsewhere.

File: verification/cbmc/sources/make_common_data_structures.c (hunk 19 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c:155]: Updated!

File: verification/cbmc/sources/make_common_data_structures.c (hunk 18 above)
Comment (tegansb) [reply verification/cbmc/sources/make_common_data_structures.c]: Great idea! Updated to use the same order as in the header struct

**Pull request merged**